### PR TITLE
fix!: various issues / cleanup + adds approve_until_date support

### LINKED ISSUE #8: approve_after_days  option is not available on approval_rule
Have a question? Please checkout our [Slack Community](https://slack.cloudposse.com) or visit our [Slack Archive](https://archive.sweetops.com/). 

[Slack Community](https://slack.cloudposse.com)

## Describe the Feature
approve_after_days  - The cutoff date for auto approval of released patches. Any patches released on or before this date are installed automatically. Date is formatted as YYYY-MM-DD

approve_after_days   option is missing in creating patch base line rule
## Use Case
approve_after_days  - can help us to approve the update within the timeline. so that we can test patches in development and we can approve to production 
Is your feature request related to a problem/challenge you are trying to solve? Please provide some additional context of why this feature or capability will be valuable.

## Describe Ideal Solution

A clear and concise description of what you want to happen. If you don't know, that's okay.

## Alternatives Considered

Explain what alternative solutions or features you've considered.

### LINKED ISSUE #14: output_s3_key_prefix variable for scan logs
## Describe the Feature
Currently we have a variable to define the output S3 key prefix for install logs as below:
output_s3_key_prefix = var.s3_bucket_prefix_install_logs

Could you please introduce a new variable for scan logs too? Currently, it is hardcoded as below:
output_s3_key_prefix = "scaning"

### LINKED ISSUE #40: Complete example Target Tagging is wrong
### Describe the Bug

Complete Example shows EC2 with Tag keys "TOSCAN" and "TOPATCH" both with values "true" and the complete example leaves all inputs as default (Very helpful to understand the 'complete' usage lol)

The Scan/Install Target Tag Key is "Patch Group", which cant be changed in this module.  So the Patch Baseline target is not looking for a "TOPATCH:true" key pair, it is looking for a "Patch Group:TOSCAN" key pair.

### Expected Behavior

Tagging an instance as "TOPATCH" = "true" would make that instance a target for a default input patch baseline.
Tagging an instance as "TOSCAN" = "true" would make that instance a target for a default input scan baseline.

### Steps to Reproduce

Try the module out.

### Screenshots

_No response_

### Environment

_No response_

### Additional Context

In the module that is credited as inspiration for this module, the key pairs can be assigned manually, allowing for much more flexibility. I would incorporate that.

### LINKED ISSUE #41: Count resolution error when providing a bucket_id
### Describe the Bug

When providing a value for `bucket_id`, module can not plan or apply.

### Expected Behavior

Ability to create my own bucket and provide it as an input

### Steps to Reproduce

```
##---------------------------------------------------
## AWS S3 Bucket for storing logs
##---------------------------------------------------
# S3 bucket for storing logs
module "log_storage" {
  source = "cloudposse/s3-log-storage/aws"
  name      = "account-guardian-logs-${data.aws_caller_identity.current.account_id}"
  namespace = "ga"
  
  s3_object_ownership           = "BucketOwnerPreferred"
  bucket_key_enabled            = true
  versioning_enabled            = true
  lifecycle_rule_enabled        = false
  lifecycle_configuration_rules = [local.lifecycle_configuration_rule]
}


##---------------------------------------------------
## AWS Systems Manager Patch Manager
##---------------------------------------------------

module "ssm-patch-manager-critical-al2" {
  source  = "cloudposse/ssm-patch-manager/aws"
  version = "0.6.0"
  region = data.aws_region.current.name

  approved_patches_compliance_level = "CRITICAL"

  namespace = "ga-account-guardian"
  operating_system = "AMAZON_LINUX_2"
  reboot_option = "NoReboot"
  scan_patch_groups = ["al2-automatic"]
  install_patch_groups = ["al2-automatic"]
  bucket_id = module.log_storage.bucket_id

  patch_baseline_approval_rules = [
    {
      approve_after_days  = 0
      compliance_level    = "HIGH"
      enable_non_security = false
      patch_baseline_filters = [
        {
          name   = "PRODUCT"
          values = ["AmazonLinux2", "AmazonLinux2.0"]
        },
        {
          name   = "CLASSIFICATION"
          values = ["Security"]
        },
        {
          name   = "SEVERITY"
          values = ["Critical", "Important"]
        }
      ]
    }
  ]

  // Schedule for the maintenance window
  // Install, and then scan to see if there are any missing patches/reboots needed
  install_maintenance_window_schedule = "cron(0 6 * * ? *)" // Run every day at 6:00 AM
  scan_maintenance_window_schedule = "cron(0 7 * * ? *)" // Run every day at 7:00 AM
}
```

### Screenshots

_No response_

### Environment

`terraform --version`
Terraform v1.4.2
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v5.37.0
+ provider registry.terraform.io/hashicorp/time v0.10.0

```
cat .terraform/modules/modules.json | jq
{
  "Modules": [
    {
      "Key": "",
      "Source": "",
      "Dir": "."
    },
    {
      "Key": "log_storage",
      "Source": "registry.terraform.io/cloudposse/s3-log-storage/aws",
      "Version": "1.4.2",
      "Dir": ".terraform/modules/log_storage"
    },
    {
      "Key": "log_storage.aws_s3_bucket",
      "Source": "registry.terraform.io/cloudposse/s3-bucket/aws",
      "Version": "3.1.2",
      "Dir": ".terraform/modules/log_storage.aws_s3_bucket"
    },
    {
      "Key": "log_storage.aws_s3_bucket.s3_user",
      "Source": "registry.terraform.io/cloudposse/iam-s3-user/aws",
      "Version": "1.1.0",
      "Dir": ".terraform/modules/log_storage.aws_s3_bucket.s3_user"
    },
    {
      "Key": "log_storage.aws_s3_bucket.s3_user.s3_user",
      "Source": "registry.terraform.io/cloudposse/iam-system-user/aws",
      "Version": "1.0.0",
      "Dir": ".terraform/modules/log_storage.aws_s3_bucket.s3_user.s3_user"
    },
    {
      "Key": "log_storage.aws_s3_bucket.s3_user.s3_user.store_write",
      "Source": "registry.terraform.io/cloudposse/ssm-parameter-store/aws",
      "Version": "0.10.0",
      "Dir": ".terraform/modules/log_storage.aws_s3_bucket.s3_user.s3_user.store_write"
    },
    {
      "Key": "log_storage.aws_s3_bucket.s3_user.s3_user.store_write.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/log_storage.aws_s3_bucket.s3_user.s3_user.store_write.this"
    },
    {
      "Key": "log_storage.aws_s3_bucket.s3_user.s3_user.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/log_storage.aws_s3_bucket.s3_user.s3_user.this"
    },
    {
      "Key": "log_storage.aws_s3_bucket.s3_user.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/log_storage.aws_s3_bucket.s3_user.this"
    },
    {
      "Key": "log_storage.aws_s3_bucket.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/log_storage.aws_s3_bucket.this"
    },
    {
      "Key": "log_storage.bucket_name",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/log_storage.bucket_name"
    },
    {
      "Key": "log_storage.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/log_storage.this"
    },
    {
      "Key": "ssm-patch-manager-critical-al2",
      "Source": "registry.terraform.io/cloudposse/ssm-patch-manager/aws",
      "Version": "0.6.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.install_window_label",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.install_window_label"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.scan_window_label",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.scan_window_label"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket",
      "Source": "registry.terraform.io/cloudposse/s3-bucket/aws",
      "Version": "4.0.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user",
      "Source": "registry.terraform.io/cloudposse/iam-s3-user/aws",
      "Version": "1.2.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.s3_user",
      "Source": "registry.terraform.io/cloudposse/iam-system-user/aws",
      "Version": "1.0.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.s3_user"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.s3_user.store_write",
      "Source": "registry.terraform.io/cloudposse/ssm-parameter-store/aws",
      "Version": "0.10.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.s3_user.store_write"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.s3_user.store_write.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.s3_user.store_write.this"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.s3_user.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.s3_user.this"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.s3_user.this"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket.this"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket_label",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.ssm_patch_log_s3_bucket_label"
    },
    {
      "Key": "ssm-patch-manager-critical-al2.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-al2.this"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20",
      "Source": "registry.terraform.io/cloudposse/ssm-patch-manager/aws",
      "Version": "0.6.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.install_window_label",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.install_window_label"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.scan_window_label",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.scan_window_label"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket",
      "Source": "registry.terraform.io/cloudposse/s3-bucket/aws",
      "Version": "4.0.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user",
      "Source": "registry.terraform.io/cloudposse/iam-s3-user/aws",
      "Version": "1.2.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.s3_user",
      "Source": "registry.terraform.io/cloudposse/iam-system-user/aws",
      "Version": "1.0.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.s3_user"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.s3_user.store_write",
      "Source": "registry.terraform.io/cloudposse/ssm-parameter-store/aws",
      "Version": "0.10.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.s3_user.store_write"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.s3_user.store_write.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.s3_user.store_write.this"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.s3_user.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.s3_user.this"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.s3_user.this"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket.this"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket_label",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.ssm_patch_log_s3_bucket_label"
    },
    {
      "Key": "ssm-patch-manager-critical-ubuntu20.this",
      "Source": "registry.terraform.io/cloudposse/label/null",
      "Version": "0.25.0",
      "Dir": ".terraform/modules/ssm-patch-manager-critical-ubuntu20.this"
    }
  ]
}
```

### Additional Context

Plan: 31 to add, 0 to change, 0 to destroy.
╷
│ Error: Invalid count argument
│ 
│   on .terraform/modules/ssm-patch-manager-critical-al2/ssm_log_bucket.tf line 19, in data "aws_iam_policy_document" "bucket_policy":
│   19:   count = local.create_log_bucket ? 1 : 0
│ 
│ The "count" value depends on resource attributes that cannot be determined until apply, so Terraform cannot predict how many
│ instances will be created. To work around this, use the -target argument to first apply only the resources that the count
│ depends on.
╵
╷
│ Error: Invalid count argument
│ 
│   on .terraform/modules/ssm-patch-manager-critical-al2/ssm_log_bucket.tf line 42, in module "ssm_patch_log_s3_bucket":
│   42:   count   = local.create_log_bucket ? 1 : 0
│ 
│ The "count" value depends on resource attributes that cannot be determined until apply, so Terraform cannot predict how many
│ instances will be created. To work around this, use the -target argument to first apply only the resources that the count
│ depends on.
╵
╷
│ Error: Invalid count argument
│ 
│   on .terraform/modules/ssm-patch-manager-critical-ubuntu20/ssm_log_bucket.tf line 19, in data "aws_iam_policy_document" "bucket_policy":
│   19:   count = local.create_log_bucket ? 1 : 0
│ 
│ The "count" value depends on resource attributes that cannot be determined until apply, so Terraform cannot predict how many
│ instances will be created. To work around this, use the -target argument to first apply only the resources that the count
│ depends on.
╵
╷
│ Error: Invalid count argument
│ 
│   on .terraform/modules/ssm-patch-manager-critical-ubuntu20/ssm_log_bucket.tf line 42, in module "ssm_patch_log_s3_bucket":
│   42:   count   = local.create_log_bucket ? 1 : 0
│ 
│ The "count" value depends on resource attributes that cannot be determined until apply, so Terraform cannot predict how many
│ instances will be created. To work around this, use the -target argument to first apply only the resources that the count
│ depends on.

### PR DESCRIPTION
**Pull request by Gowiem**
## what

* Converts `bucket_id` variable to `list(string)` 
* Removes `region` variable
* Cleans up some potential `enabled = false` issues in `outputs.tf` they were dependent on resources existing
* Fixes typo in scan logs bucket prefix
* Adds `s3_bucket_prefix_scan_logs` variable 
* Adds `approve_until_date` support to `patch_baseline_approval_rules` variable 

## why

* #41 shows that passing in a `bucket_id` results in a known issue: https://github.com/hashicorp/terraform/issues/28962. [We've seen this internally before](https://github.com/cloudposse/terraform-aws-security-group/pull/17) and our suggested pattern is to switch to `list(string)` as the type and check the `length`. 
	* (BREAKING CHANGE)
* Features
	* Adds support for passing a custom prefix for scan logs.
	* Adds support the ability to patch on a timeline via `approve_until_date`
* General cleanup of issues that I saw for this module:
	* region variable not used, so needed to be removed (BREAKING CHANGE)
	* Bumps s3-module to latest
	* Typo
	* Updates example to an actual use-case where patching would occur
	* Removes an unnecessary label
	* Avoids issues with outputs not working in the case that a user passed `enabled = false` since outputs were dependent on resources being at position 0 via list access. 

## references

* Closes #41
* Closes #39
* Closes #8
* Closes #40 
* Closes #14

### CODE REVIEW

File: examples/complete/main.tf [hunk 1]
```diff
@@ -37,16 +37,15 @@ module "ec2_instance" {
   ssm_patch_manager_s3_log_bucket = format("%s-%s-%s-%s", module.this.namespace, module.this.environment, module.this.stage, module.this.name)
 
   tags = {
-    "TOSCAN"  = "true",
-    "TOPATCH" = "true"
+    "Patch Group": "TOPATCH",
```
[the remark this answers is not in the document]
Comment (Gowiem) [reply examples/complete/main.tf]: Wow one damn space is what got me. 😭😭😭😭

**Review by aknysh — APPROVED** [review]
LGTM, thanks @Gowiem

**Review by jamengual — APPROVED** [review]
LGTM

**Pull request merged**